### PR TITLE
JavaScript: add string concatenation library

### PR DESCRIPTION
**Pull request by asger-semmle**
Adds a very trimmed-down version of the string concatenation library from earlier this year.

The original version was more general in some ways, but at the cost of having more intermediate predicates, some of which became materialized. The cost of materializing these intermediate predicates was linear but still became significant due to the large number string concatenations and `+` expressions in an average project.

The leaf-traversal feature and array tracking have been removed. We can consider adding them as a new feature in a later PR.

The changes that could affect results, such as including `array.join()` as a string concatenation (and therefore a taint step), have already landed separately. I believe the only concrete difference remaining is that sanitizing prefixes are now recognized in more cases. I haven't observed any changes to the results.

Performance now looks [good](https://git.semmle.com/gist/asger/3840bdc889b8f530caf8bebd44f44189) and results are the same.

### CODE REVIEW

File: javascript/ql/src/semmle/javascript/StringConcatenation.qll [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+  /** Gets the last operand to the string concatenation defining `node`. */
+  DataFlow::Node getLastOperand(DataFlow::Node node) {
+    result = getOperand(node, getNumOperand(node) - 1)
+  }
+  
+  /**
+   * Holds if `src` flows to `dst` through the `n`th operand of the given concatenation operator.
+   */
+  predicate taintStep(DataFlow::Node src, DataFlow::Node dst, DataFlow::Node operator, int n) {
```
Comment (esben-semmle) [inline javascript/ql/src/semmle/javascript/StringConcatenation.qll]: I looks a bit weird to have both `operator` and `dst` here, when `operator = dst`. But looking at the use-site, it makes perfect sense to have both, good call.

Comment (asger-semmle) [reply javascript/ql/src/semmle/javascript/StringConcatenation.qll]: And it's forward compatible with handling of `push` on arrays, where `dst != operator` is a possibility (`operator` being the `push` call itself).

**Comment by asger-semmle** [comment]
> This needs a change note, I propose to generalize the existing bullet for arrays:

Thinking about this I just uncovered a bug in the library, which I'm pushing now.

For an expression like `[a,b].join("")` the ordinary heap taint steps generate this path:
```
b => [a,b] => [a,b].join("")
```
But the `StringConcatenation::taintStep` adds the direct step:
```
b => [a,b].join("")
```
This mismatch means that if `a` is a sanitizing prefix, the direct step will be sanitized, but taint can still propagate along the longer path through the array literal.

The reason for exposing `StringConcatenation::taintStep` was so users of the library wouldn't be bothered with details like that, and it was supposed to handle this.

So I've changed how `[a,b].join()` is modelled in the library. The array literal `[a,b]` is now a concatenation operator with `a,b` as operands, and the `join` call itself is a concatenation with `[a,b]` as its only operand.

Now about the change note: I'd argue that the only difference is that sanitizing prefixes are recognized in `[a,b].join("")`, which already counts as flow through an array.

**Comment by asger-semmle** [comment]
I've started another evaluation

**Comment by asger-semmle** [comment]
[Updated evaluation](https://git.semmle.com/asger/dist-compare-reports/tree/f0bc6df61ad8c2ee2257c71c70f43f85). Same story.

**Comment by esben-semmle** [comment]
(minor expected output mismatch)

**Pull request merged**